---
title: "Figure 4"
author: "Huang Lin"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
rm(list = ls())

library(tidyverse)
library(nloptr)
library(phyloseq)
library(stringr)
library(RColorBrewer)
library(ggpubr)
library(ggforce)
library(limma)
library(vegan)
library(MASS)
library(pander)
panderOptions('table.caption.prefix', NULL)
panderOptions('table.continues', NULL)
panderOptions('table.emphasize.rownames', FALSE)

source("ancom.R")
```

# 1. Read in OTU table and meta data

```{r, message=FALSE, warning=FALSE, comment=NA}
# Meta data
meta_data=read_tsv("../data/global_gut/global_gut_metadata.txt")
meta_data=meta_data%>%transmute(Sample.ID=`#SampleID`, age=AGE, sex=SEX, country=COUNTRY)%>%
  arrange(Sample.ID)
meta_data=meta_data[complete.cases(meta_data), ]
meta_data$age=as.numeric(meta_data$age)
meta_data$country=recode(meta_data$country, `GAZ:Malawi` = "MA", 
                         `GAZ:United States of America` = "US", `GAZ:Venezuela` = "VEN")

# Taxonomy
taxonomy=read_tsv("../data/global_gut/global_gut_taxonomy.txt")
taxonomy=taxonomy%>%rowwise()%>%
  mutate(genus_name=paste(Phylum, Genus, sep = ";"))

# OTU table
otu_table=read_tsv("../data/global_gut/global_gut_otu.txt")
otu_table=otu_table[, -532]
otu_table$OTU_ID=taxonomy$genus_name[match(otu_table$OTU_ID, taxonomy$OTU_ID)]
otu_table=as.data.frame(otu_table)
otu_table[, -1]=apply(otu_table[, -1], 2, as.numeric)

# Aggregate into genus level
genus_table=otu_table%>%group_by(OTU_ID)%>%
  summarise_all(sum)
non_info_pos=grep("\\g__\\b", genus_table$OTU_ID) # Exact match
genus_table=genus_table[-non_info_pos, ]
genus_table=as.data.frame(genus_table)
```

# 2. Data Pre-Processing

```{r, message=FALSE, warning=FALSE, comment=NA}
feature_table_pre_process=function(feature.table, meta.data, sample.var, group.var, pre.cut, neg.lb){
  feature.table=as.data.frame(feature.table)
  meta.data=as.data.frame(meta.data)
  meta.data[]=lapply(meta.data, function(x) if(is.factor(x)) factor(x) else x)
  
  sample.ID=colnames(feature.table)
  meta.data=meta.data[match(sample.ID, meta.data[, sample.var]), ]
  
  # 1. Identify outliers within each taxon
  group=as.factor(meta.data[, group.var])
  group.name=levels(group)
  grp.ind.origin=lapply(1:nlevels(group), function(i) which(group==group.name[i]))
  n.grp.origin=length(grp.ind.origin)
  n.samp.grp.origin=sapply(grp.ind.origin, length)
  feature.table=feature.table[, unlist(grp.ind.origin)]

  z=log(feature.table+1)
  f=z; f[f==0]=NA; f=colMeans(f, na.rm = T)
  f.mean=unlist(tapply(f, rep(1:n.grp.origin, n.samp.grp.origin), mean))
  e=f-rep(f.mean, n.samp.grp.origin)
  y=t(t(z)-e)
  
  outlier_check=function(x){
    mu1=quantile(x, 0.25); mu2=quantile(x, 0.75)
    sigma1=quantile(x, 0.75)-quantile(x, 0.25); sigma2=sigma1
    pi=0.75
    n=length(x)
    epsilon=100
    tol=1e-5
    score=pi*dnorm(x, mean = mu1, sd=sigma1)/((1-pi)*dnorm(x, mean = mu2, sd=sigma2))
    while (epsilon>tol) {
      grp1.ind=score>=1
      mu1.new=mean(x[grp1.ind]); mu2.new=mean(x[!grp1.ind])
      sigma1.new=sd(x[grp1.ind]); if(is.na(sigma1.new)) sigma1.new=0
      sigma2.new=sd(x[!grp1.ind]); if(is.na(sigma2.new)) sigma2.new=0
      pi.new=sum(grp1.ind)/n
      
      para=c(mu1.new, mu2.new, sigma1.new, sigma2.new, pi.new)
      if(any(is.na(para))) break
      
      score=pi.new*dnorm(x, mean = mu1.new, sd=sigma1.new)/
        ((1-pi.new)*dnorm(x, mean = mu2.new, sd=sigma2.new))
      
      epsilon=sqrt((mu1-mu1.new)^2+(mu2-mu2.new)^2+
                     (sigma1-sigma1.new)^2+(sigma2-sigma2.new)^2+(pi-pi.new)^2)
      mu1=mu1.new; mu2=mu2.new; sigma1=sigma1.new; sigma2=sigma2.new; pi=pi.new
    }
    
    if(mu1+1.96*sigma1<mu2-1.96*sigma2){
      if(pi>0.85){
        out.ind=(!grp1.ind)
      }else if(pi<0.15){
        out.ind=grp1.ind
      }else{
        out.ind=rep(FALSE, n)
      }
    }else{
      out.ind=rep(FALSE, n)
    }
    return(out.ind)
  }
  feature.table.out=t(apply(y, 1, function(i)
    unlist(tapply(i, rep(1:n.grp.origin, n.samp.grp.origin), function(j) outlier_check(j)))))
  feature.table[feature.table.out]=NA
  
  # 2. Filter out taxa with zeros >= pre.cut
  taxa.zero.prop=apply(feature.table, 1, function(x) sum(x==0, na.rm = T)/length(x[!is.na(x)]))
  filter.taxa=which(taxa.zero.prop>=pre.cut)
  if(length(filter.taxa)>0){
    feature.table=feature.table[-filter.taxa, ]
  }
  
  # 3. Filter out subjects with library size < 1000
  library.size=colSums(feature.table, na.rm = T)
  sample.ID=colnames(feature.table)
  meta.data=meta.data[match(sample.ID, meta.data[, sample.var]), ]
  
  if(any(library.size<1000)){
    filter.subject=which(library.size<1000)
    feature.table=feature.table[, -filter.subject]
    meta.data=meta.data[-filter.subject, ]
  }
  
  # 4. Re-order the OTU table
  group=as.factor(meta.data[, group.var])
  group.name=levels(group)
  grp.ind=lapply(1:nlevels(group), function(i) which(group==group.name[i]))
  n.grp=length(grp.ind)
  n.samp.grp=sapply(grp.ind, length)
  
  n.taxa=nrow(feature.table)
  taxa.id=rownames(feature.table)
  n.samp=ncol(feature.table)
  
  # 5. Identify taxa containing structure zeros
  present.table=as.matrix(feature.table)
  present.table[is.na(present.table)]=0
  present.table[present.table!=0]=1
  
  p.hat.mat=t(apply(present.table, 1, function(x)
    unlist(tapply(x, rep(1:n.grp, n.samp.grp), function(y) mean(y, na.rm = T)))))
  sample.size=t(apply(feature.table, 1, function(x)
    unlist(tapply(x, rep(1:n.grp, n.samp.grp), function(y) length(y[!is.na(y)])))))
  p.hat.lo.mat=p.hat.mat-1.96*sqrt(p.hat.mat*(1-p.hat.mat)/sample.size)
  colnames(p.hat.mat)=levels(group)
  colnames(p.hat.lo.mat)=levels(group)
  
  struc.zero=matrix(0, nrow = n.taxa, ncol = n.grp)
  struc.zero[p.hat.mat==0]=1
  # Whether we need to classify a taxon into structural zero by its negative lower bound?
  if(neg.lb) struc.zero[p.hat.lo.mat<=0]=1
  rownames(struc.zero)=taxa.id
  colnames(struc.zero)=paste0("Structural Zero in ", levels(group))
  
  # 6. Return results
  res=list(feature.table=feature.table, library.size=library.size, 
           group.name=group.name, group.ind=grp.ind, structure.zeros=struc.zero)
  return(res)
}
```

# 3. ANCOM-BC

```{r, message=FALSE, warning=FALSE, comment=NA}
ANCOM_BC=function(feature.table, grp.name, grp.ind, struc.zero, adj.method, 
                  tol.EM, max.iterNum, alpha){
  n.taxa.origin=nrow(feature.table)
  taxa.id.origin=rownames(feature.table)
  n.samp=ncol(feature.table)
  sample.id=colnames(feature.table)
  
  n.grp=length(grp.ind)
  n.samp.grp=sapply(grp.ind, length)
  
  ### 0. Separate out taxa with no structural zeros
  info.taxa.pos=which(apply(struc.zero, 1, function(x) all(x==0)))
  O=feature.table[info.taxa.pos, ]
  n.taxa=nrow(O)
  taxa.id=rownames(O)
  n.samp=ncol(O)
  y=log(O+1)
  
  ### 1. Parameters estimates
  d=rep(0, n.samp)
  mu=t(apply(y, 1, function(i) tapply(i, rep(1:n.grp, n.samp.grp), function(j)
    mean(j, na.rm = T))))
  iterNum=0
  epsilon=100
  while (epsilon>tol.EM&iterNum<max.iterNum) {
    # Updating mu
    mu.new=t(apply(t(t(y)-d), 1, function(i) tapply(i, rep(1:n.grp, n.samp.grp), function(j)
      mean(j, na.rm = T))))
    
    # Updating d
    d.new=colMeans(y-mu.new[, rep(1:ncol(mu.new), times = n.samp.grp)], na.rm = T)
    
    # Iteration
    epsilon=sqrt(sum((mu.new-mu)^2)+sum((d.new-d)^2))
    iterNum=iterNum+1
    
    mu=mu.new
    d=d.new
  }
  
  mu.var.raw=(y-t(t(mu[, rep(1:ncol(mu), times = n.samp.grp)])+d))^2
  mu.var=t(apply(mu.var.raw, 1, function(x) tapply(x, rep(1:n.grp, n.samp.grp), function(y)
    mean(y, na.rm = T))))
  sample.size=t(apply(y, 1, function(x)
    unlist(tapply(x, rep(1:n.grp, n.samp.grp), function(y) length(y[!is.na(y)])))))
  mu.var=mu.var/sample.size
  
  ### 2. Estimate the between-group by E-M algorithm
  Delta=mu[, 1]-mu[, 2]
  Delta.var.est=rowSums(mu.var)
  
  ## 2.1 Initials
  pi1_0=0.125
  pi2_0=0.75
  pi3_0=0.125
  delta_0=mean(Delta[Delta>=quantile(Delta, 0.25, na.rm = T)&
                       Delta<=quantile(Delta, 0.75, na.rm = T)], na.rm = T)
  d1_0=mean(Delta[Delta<quantile(Delta, 0.125, na.rm = T)], na.rm = T)
  d2_0=mean(Delta[Delta>quantile(Delta, 0.875, na.rm = T)], na.rm = T)
  psi1.sq_0=var(Delta[Delta<quantile(Delta, 0.125, na.rm = T)], na.rm = T)
  if(is.na(psi1.sq_0)|psi1.sq_0==0) psi1.sq_0=1
  psi2.sq_0=var(Delta[Delta>quantile(Delta, 0.875, na.rm = T)], na.rm = T)
  if(is.na(psi2.sq_0)|psi2.sq_0==0) psi2.sq_0=1
  
  ## 2.2 Apply E-M algorithm
  # 2.21 Store all paras in vectors/matrices
  pi1.vec=c(pi1_0); pi2.vec=c(pi2_0); pi3.vec=c(pi3_0)
  delta.vec=c(delta_0); d1.vec=c(d1_0); d2.vec=c(d2_0)
  psi1.sq.vec=c(psi1.sq_0); psi2.sq.vec=c(psi2.sq_0)
  
  # 2.22 E-M iteration
  iterNum=0
  epsilon=100
  sigmai.sq=Delta.var.est
  while (epsilon>tol.EM&iterNum<max.iterNum) {
    # print(iterNum)
    ## Current value of paras
    pi1=pi1.vec[length(pi1.vec)]; pi2=pi2.vec[length(pi2.vec)]; pi3=pi3.vec[length(pi3.vec)]
    delta=delta.vec[length(delta.vec)]; 
    d1=d1.vec[length(d1.vec)]; d2=d2.vec[length(d2.vec)]
    psi1.sq=psi1.sq.vec[length(psi1.sq.vec)]; psi2.sq=psi2.sq.vec[length(psi2.sq.vec)]
    
    ## E-step
    pdf1=sapply(seq(n.taxa), function(i) dnorm(Delta[i], delta+d1, sqrt(sigmai.sq[i]+psi1.sq)))
    pdf2=sapply(seq(n.taxa), function(i) dnorm(Delta[i], delta, sqrt(sigmai.sq[i])))
    pdf3=sapply(seq(n.taxa), function(i) dnorm(Delta[i], delta+d2, sqrt(sigmai.sq[i]+psi2.sq)))
    r1i=pi1*pdf1/(pi1*pdf1+pi2*pdf2+pi3*pdf3); r1i[is.na(r1i)]=0
    r2i=pi2*pdf2/(pi1*pdf1+pi2*pdf2+pi3*pdf3); r2i[is.na(r2i)]=0
    r3i=pi3*pdf3/(pi1*pdf1+pi2*pdf2+pi3*pdf3); r3i[is.na(r3i)]=0
    
    ## M-step
    pi1_new=mean(r1i, na.rm = T); pi2_new=mean(r2i, na.rm = T); pi3_new=mean(r3i, na.rm = T)
    delta_new=sum(r1i*(Delta-d1)/(sigmai.sq+psi1.sq)+r2i*Delta/sigmai.sq+
                    r3i*(Delta-d2)/(sigmai.sq+psi2.sq), na.rm = T)/
      sum(r1i/(sigmai.sq+psi1.sq)+r2i/sigmai.sq+r3i/(sigmai.sq+psi2.sq), na.rm = T)
    d1_new=min(sum(r1i*(Delta-delta)/(sigmai.sq+psi1.sq), na.rm = T)/
                 sum(r1i/(sigmai.sq+psi1.sq), na.rm = T), 0)
    if(is.nan(d1_new)) d1_new=0
    d2_new=max(sum(r3i*(Delta-delta)/(sigmai.sq+psi2.sq), na.rm = T)/
                 sum(r3i/(sigmai.sq+psi2.sq), na.rm = T), 0)
    if(is.nan(d2_new)) d2_new=0
    
    # Nelder-Mead simplex algorithm for psi1.sq, psi2.sq, and sigmai.sq
    obj.psi1.sq=function(x){
      log.pdf=log(sapply(seq(n.taxa), function(i) dnorm(Delta[i], delta+d1, sqrt(sigmai.sq[i]+x))))
      log.pdf[is.infinite(log.pdf)]=0
      -sum(r1i*log.pdf, na.rm = T)
    }
    psi1.sq_new=neldermead(x0 = psi1.sq, fn = obj.psi1.sq, lower = 0)$par
    
    obj.psi2.sq=function(x){
      log.pdf=log(sapply(seq(n.taxa), function(i) dnorm(Delta[i], delta+d2, sqrt(sigmai.sq[i]+x))))
      log.pdf[is.infinite(log.pdf)]=0
      -sum(r3i*log.pdf, na.rm = T)
    }
    psi2.sq_new=neldermead(x0 = psi2.sq, fn = obj.psi2.sq, lower = 0)$par
    
    ## Merge to the paras vectors/matrices
    pi1.vec=c(pi1.vec, pi1_new); pi2.vec=c(pi2.vec, pi2_new); pi3.vec=c(pi3.vec, pi3_new)
    delta.vec=c(delta.vec, delta_new)
    d1.vec=c(d1.vec, d1_new); d2.vec=c(d2.vec, d2_new)
    psi1.sq.vec=c(psi1.sq.vec, psi1.sq_new); psi2.sq.vec=c(psi2.sq.vec, psi2.sq_new)
    
    ## Calculate the new epsilon
    epsilon=sqrt((pi1_new-pi1)^2+(pi2_new-pi2)^2+(pi3_new-pi3)^2+(delta_new-delta)^2+
                   (d1_new-d1)^2+(d2_new-d2)^2+(psi1.sq_new-psi1.sq)^2+(psi2.sq_new-psi2.sq)^2)
    iterNum=iterNum+1
  }
  bias.est=delta.vec[length(delta.vec)]
  
  ### 4. Test results
  ## 4.1 Results for taxa with non-structural zeros
  W.numerator=mu[, 1]-mu[, 2]-bias.est
  W.denominator=mu.var[, 1]+mu.var[, 2]
  
  W=W.numerator/sqrt(W.denominator)
  p.val=sapply(W, function(x) 2*pnorm(abs(x), mean=0, sd=1, lower.tail = F))
  q.val=p.adjust(p.val, method = adj.method)
  q.val[is.na(q.val)]=1
  
  res.nonstrc.zero=data.frame(W.numerator, se=sqrt(W.denominator), W, p.val, q.val)
  
  ## 4.2 Results for taxa with structural zeros
  if(length(info.taxa.pos)<n.taxa.origin){
    O.strc.zero=feature.table[-info.taxa.pos, ]
    ind.strc.zero=struc.zero[-info.taxa.pos, rep(1:n.grp, times = n.samp.grp)]
    O.strc.zero.adj=O.strc.zero*(1-ind.strc.zero)
    y.strc.zero=log(O.strc.zero.adj+1)
    d.strc.zero=t(t(1-ind.strc.zero)*d)
    y.strc.zero.adj=y.strc.zero-d.strc.zero
    mu.strc.zero=t(apply(y.strc.zero.adj, 1, function(i) 
      tapply(i, rep(1:n.grp, n.samp.grp), function(j) mean(j, na.rm = T))))
    # Make it the relative mean difference (with related to the smallest value)
    mu.strc.zero.adj=mu.strc.zero
    mu.strc.zero.adj[mu.strc.zero.adj==0]=NA
    mu.strc.zero.adj=t(t(mu.strc.zero.adj)+abs(apply(mu.strc.zero, 2, min)))
    mu.strc.zero.adj[is.na(mu.strc.zero.adj)]=0
    
    res.strc.zero=data.frame(W.numerator=mu.strc.zero.adj[, 1]-mu.strc.zero.adj[, 2], 
                             se=0, W=Inf, p.val=0, q.val=0)
  }else{
    res.strc.zero=NA
  }
  
  ## 4.3 Combine these results together
  res=data.frame(W.numerator=Inf, se=0, W=Inf, 
                 p.val=rep(0, n.taxa.origin), q.val=rep(0, n.taxa.origin))
  res[info.taxa.pos, ]=res.nonstrc.zero
  res[-info.taxa.pos, ]=res.strc.zero
  res=mutate(res, diff.abn=ifelse(q.val<alpha, TRUE, FALSE))
  
  colnames(res)=c(paste0("mean.difference (", grp.name[1], " - ", grp.name[2], ")"), 
                  "se", paste0("effect.size (", grp.name[1], " - ", grp.name[2], ")"), 
                  "p.val", "q.val", "diff.abn")
  out=list(feature.table=feature.table, res=res, d=d, mu=mu, bias.est=bias.est)
  return(out)
}
```

# 4. MA vs US

## 4.1 Subset OTU table and meta data

```{r, message=FALSE, warning=FALSE, comment=NA}
# Subset meta data
meta.data=meta_data%>%filter(country%in%c("MA", "US"))
meta.data$country=as.character(meta.data$country)
# Subset OTU table
obs.abn=genus_table
rownames(obs.abn)=obs.abn$OTU_ID
obs.abn=obs.abn[, -1]
obs.abn=obs.abn[, meta.data$Sample.ID]

# Recode sample ID
meta.data$Sample.ID=seq(nrow(meta.data))
colnames(obs.abn)=seq(nrow(meta.data))

write_tsv(meta.data, "../data/songbird/meta_ma_us.txt")
write_tsv(obs.abn%>%rownames_to_column("genus"), "../data/songbird/otu_ma_us.txt")
```

## 4.2 Pre-Processing

```{r, message=FALSE, warning=FALSE, comment=NA}
feature.table=obs.abn; sample.var="Sample.ID"; group.var="country"; pre.cut=0.90; neg.lb=TRUE
pre.process=feature_table_pre_process(feature.table, meta.data, sample.var, group.var, pre.cut, neg.lb)
feature.table=pre.process$feature.table
library.size=pre.process$library.size
group.name=pre.process$group.name
group.ind=pre.process$group.ind
struc.zero=pre.process$structure.zeros
num.struc.zero=apply(struc.zero, 1, sum)
s0=rownames(feature.table)[which(num.struc.zero==0)]
s1=rownames(feature.table)[which(num.struc.zero==1)]
```

## 4.3 ANCOM-BC

```{r, message=FALSE, warning=FALSE, comment=NA, eval=FALSE}
# Paras for ANCOM-BC
grp.name=group.name; grp.ind=group.ind; adj.method="bonferroni"
tol.EM=1e-5; max.iterNum=100; alpha=0.05

# Run ANCOM-BC
start_time <- Sys.time()
out=ANCOM_BC(feature.table, grp.name, grp.ind, struc.zero, adj.method, 
              tol.EM, max.iterNum, alpha)
end_time <- Sys.time()
end_time - start_time

# Test result
res.ANCOM_BC=data.frame(genus=rownames(out$feature.table), out$res, 
                        struc.zero[rownames(out$feature.table), ], 
                        row.names = NULL, stringsAsFactors=FALSE)
res.ANCOM_BC$phylum=sapply(res.ANCOM_BC$genus, function(x) strsplit(x, ".g")[[1]][1])
res.ANCOM_BC=res.ANCOM_BC%>%transmute(phylum, genus,
                                      `mean.difference (MA - US)` = mean.difference..MA...US.,
                                      se, `effect.size (MA - US)`=effect.size..MA...US., 
                                      p.val, q.val, diff.abn, 
                                      struc.zero.MA=Structural.Zero.in.MA, 
                                      struc.zero.US=Structural.Zero.in.US)
res.ANCOM_BC=res.ANCOM_BC%>%arrange(q.val)
res.ANCOM_BC.s0=filter(res.ANCOM_BC, genus%in%s0)

write.csv(res.ANCOM_BC, "Output/Real ../data/Genus Level/Taxa_MA vs US/MA vs US_ANCOM-BC_genus_level.csv", row.names = F)
```

## 4.4 ANCOM

```{r, message=FALSE, warning=FALSE, comment=NA, eval=FALSE}
# Run ANCOM
otu.data=t(obs.abn[s0, ])
otu.data=data.frame(Sample.ID=colnames(obs.abn), otu.data, stringsAsFactors=F)
colnames(otu.data)=c("Sample.ID", s0)

start_time <- Sys.time()
res.W=ANCOM.main(OTUdat=otu.data, Vardat=meta.data, adjusted=FALSE, repeated=FALSE,
                 main.var="country", adj.formula=NULL, repeat.var=NULL, longitudinal=FALSE,
                 random.formula=NULL, multcorr=2, sig=0.05, prev.cut=1.01)

end_time <- Sys.time()
end_time - start_time

res.ANCOM=data.frame(otu.names=c(s0, s1), W_stat=Inf, detected_0.9=TRUE,
                     detected_0.8=TRUE, detected_0.7=TRUE, detected_0.6=TRUE)
res.ANCOM[match(res.W$W.taxa$otu.names, res.ANCOM$otu.names), ]=res.W$W.taxa
res.ANCOM=res.ANCOM%>%arrange(desc(detected_0.7))
res.ANCOM.s0=res.ANCOM[match(s0, res.ANCOM$otu.names), ]

write.csv(res.ANCOM, "Output/Real ../data/Genus Level/Taxa_MA vs US/MA vs US_ANCOM_genus_level.csv", row.names = F)
```

## 4.5 Venn diagram

```{r, message=FALSE, warning=FALSE, comment=NA}
res.ANCOM_BC = read_csv("../data/global_gut/ma_us_ancom_bc_genus.csv")
res.ANCOM = read_csv("../data/global_gut/ma_us_ancom_genus.csv")
res.DR = read_tsv("../data/global_gut/ma_us_dr_genus.tsv")
res.DR = res.DR%>%arrange(desc(`country[T.US]`))
# Tabulation
sig.taxa = data.frame(ANCOM_BC = rownames(obs.abn)%in%res.ANCOM_BC[res.ANCOM_BC$diff.abn, ]$genus*1,
                      ANCOM = rownames(obs.abn)%in%res.ANCOM[res.ANCOM$detected_0.7, ]$otu.names*1,
                      DR = rownames(obs.abn)%in%res.DR[c(1:25, (nrow(res.DR)-24):nrow(res.DR)), ]$X1*1)
rownames(sig.taxa) = rownames(obs.abn)
venn.ma_us = vennCounts(sig.taxa)

vennDiagram(venn.ma_us, circle.col = c("red", "blue", "green3"), 
            names = c("ANCOM-BC", "ANCOM", "DR"))
```

# 5. VEN vs US

## 5.1 Subset OTU table and meta data

```{r, message=FALSE, warning=FALSE, comment=NA}
# Subset meta data
meta.data=meta_data%>%filter(country%in%c("VEN", "US"))
meta.data$country=as.character(meta.data$country)
# Subset OTU table
obs.abn=genus_table
rownames(obs.abn)=obs.abn$OTU_ID
obs.abn=obs.abn[, -1]
obs.abn=obs.abn[, meta.data$Sample.ID]

# Recode sample ID
meta.data$Sample.ID=seq(nrow(meta.data))
colnames(obs.abn)=seq(nrow(meta.data))

write_tsv(meta.data, "../data/songbird/meta_ven_us.txt")
write_tsv(obs.abn%>%rownames_to_column("genus"), "../data/songbird/otu_ven_us.txt")
```

## 5.2 Pre-Processing

```{r, message=FALSE, warning=FALSE, comment=NA}
feature.table=obs.abn; sample.var="Sample.ID"; group.var="country"; pre.cut=0.90; neg.lb=TRUE
pre.process=feature_table_pre_process(feature.table, meta.data, sample.var, group.var, pre.cut, neg.lb)
feature.table=pre.process$feature.table
library.size=pre.process$library.size
group.name=pre.process$group.name
group.ind=pre.process$group.ind
struc.zero=pre.process$structure.zeros
num.struc.zero=apply(struc.zero, 1, sum)
s0=rownames(feature.table)[which(num.struc.zero==0)]
s1=rownames(feature.table)[which(num.struc.zero==1)]
```

## 5.3 ANCOM-BC

```{r, message=FALSE, warning=FALSE, comment=NA, eval=FALSE}
# Paras for ANCOM-BC
grp.name=group.name; grp.ind=group.ind; adj.method="bonferroni"
tol.EM=1e-5; max.iterNum=100; alpha=0.05

# Run ANCOM-BC
start_time <- Sys.time()
out=ANCOM_BC(feature.table, grp.name, grp.ind, struc.zero, adj.method, 
              tol.EM, max.iterNum, alpha)
end_time <- Sys.time()
end_time - start_time

# Test result
res.ANCOM_BC=data.frame(genus=rownames(out$feature.table), out$res, 
                        struc.zero[rownames(out$feature.table), ], 
                        row.names = NULL, stringsAsFactors=FALSE)
res.ANCOM_BC$phylum=sapply(res.ANCOM_BC$genus, function(x) strsplit(x, ".g")[[1]][1])
res.ANCOM_BC=res.ANCOM_BC%>%transmute(phylum, genus,
                                      `mean.difference (VEN - US)` = -mean.difference..US...VEN.,
                                      se, `effect.size (VEN - US)` = ifelse(is.infinite(effect.size..US...VEN.), Inf, -effect.size..US...VEN.), 
                                      p.val, q.val, diff.abn, 
                                      struc.zero.VEN=Structural.Zero.in.VEN, 
                                      struc.zero.US=Structural.Zero.in.US)
res.ANCOM_BC=res.ANCOM_BC%>%arrange(q.val)
res.ANCOM_BC.s0=filter(res.ANCOM_BC, genus%in%s0)

write.csv(res.ANCOM_BC, "Output/Real ../data/Genus Level/Taxa_VEN vs US/VEN vs US_ANCOM-BC_genus_level.csv", row.names = F)
```

## 5.4 ANCOM

```{r, message=FALSE, warning=FALSE, comment=NA, eval=FALSE}
# Run ANCOM
otu.data=t(obs.abn[s0, ])
otu.data=data.frame(Sample.ID=colnames(obs.abn), otu.data, stringsAsFactors=F)
colnames(otu.data)=c("Sample.ID", s0)

start_time <- Sys.time()
res.W=ANCOM.main(OTUdat=otu.data, Vardat=meta.data, adjusted=FALSE, repeated=F,
                 main.var="country", adj.formula=NULL, repeat.var=NULL, longitudinal=FALSE,
                 random.formula=NULL, multcorr=2, sig=0.05, prev.cut=1.01)

end_time <- Sys.time()
end_time - start_time

res.ANCOM=data.frame(otu.names=c(s0, s1), W_stat=Inf, detected_0.9=TRUE,
                     detected_0.8=TRUE, detected_0.7=TRUE, detected_0.6=TRUE)
res.ANCOM[match(res.W$W.taxa$otu.names, res.ANCOM$otu.names), ]=res.W$W.taxa
res.ANCOM=res.ANCOM%>%arrange(desc(detected_0.7))
res.ANCOM.s0=res.ANCOM[match(s0, res.ANCOM$otu.names), ]

# Tabulation
# res.compare=cbind(select(res.ANCOM_BC.s0, genus, diff.abn), W=res.ANCOM.s0$detected_0.7)
res.compare=left_join(res.ANCOM_BC, res.ANCOM,  by = c("genus" = "otu.names"))
tab.ven.us=xtabs(~ diff.abn + detected_0.7, data=res.compare)
tab.ven.us=addmargins(tab.ven.us)
pander(tab.ven.us)

write.csv(res.ANCOM, "Output/Real ../data/Genus Level/Taxa_VEN vs US/VEN vs US_ANCOM_genus_level.csv", row.names = F)
```

## 5.5 Venn diagram

```{r, message=FALSE, warning=FALSE, comment=NA}
res.ANCOM_BC = read_csv("../data/global_gut/ven_us_ancom_bc_genus.csv")
res.ANCOM = read_csv("../data/global_gut/ven_us_ancom_genus.csv")
res.DR = read_tsv("../data/global_gut/ven_us_dr_genus.tsv")
res.DR = res.DR%>%arrange(desc(`country[T.VEN]`))
# Tabulation
sig.taxa = data.frame(ANCOM_BC = rownames(obs.abn)%in%res.ANCOM_BC[res.ANCOM_BC$diff.abn, ]$genus*1,
                      ANCOM = rownames(obs.abn)%in%res.ANCOM[res.ANCOM$detected_0.7, ]$otu.names*1,
                      DR = rownames(obs.abn)%in%res.DR[c(1:25, (nrow(res.DR)-24):nrow(res.DR)), ]$X1*1)
rownames(sig.taxa) = rownames(obs.abn)
venn.ven_us = vennCounts(sig.taxa)

vennDiagram(venn.ven_us, circle.col = c("red", "blue", "green3"), 
            names = c("ANCOM-BC", "ANCOM", "DR"))
```

# 6. MA vs VEN

## 6.1 Subset OTU table and meta data

```{r, message=FALSE, warning=FALSE, comment=NA}
# Subset meta data
meta.data=meta_data%>%filter(country%in%c("MA", "VEN"))
meta.data$country=as.character(meta.data$country)
# Subset OTU table
obs.abn=genus_table
rownames(obs.abn)=obs.abn$OTU_ID
obs.abn=obs.abn[, -1]
obs.abn=obs.abn[, meta.data$Sample.ID]

# Recode sample ID
meta.data$Sample.ID=seq(nrow(meta.data))
colnames(obs.abn)=seq(nrow(meta.data))

write_tsv(meta.data, "../data/songbird/meta_ma_ven.txt")
write_tsv(obs.abn%>%rownames_to_column("genus"), "../data/songbird/otu_ma_ven.txt")
```

## 6.2 Pre-Processing

```{r, message=FALSE, warning=FALSE, comment=NA}
feature.table=obs.abn; sample.var="Sample.ID"; group.var="country"; pre.cut=0.90; neg.lb=TRUE
pre.process=feature_table_pre_process(feature.table, meta.data, sample.var, group.var, pre.cut, neg.lb)
feature.table=pre.process$feature.table
library.size=pre.process$library.size
group.name=pre.process$group.name
group.ind=pre.process$group.ind
struc.zero=pre.process$structure.zeros
num.struc.zero=apply(struc.zero, 1, sum)
s0=rownames(feature.table)[which(num.struc.zero==0)]
s1=rownames(feature.table)[which(num.struc.zero==1)]
```

## 6.3 ANCOM-BC

```{r, message=FALSE, warning=FALSE, comment=NA, eval=FALSE}
# Paras for ANCOM-BC
grp.name=group.name; grp.ind=group.ind; adj.method="bonferroni"
tol.EM=1e-5; max.iterNum=100; alpha=0.05

# Run ANCOM-BC
start_time <- Sys.time()
out=ANCOM_BC(feature.table, grp.name, grp.ind, struc.zero, adj.method, 
              tol.EM, max.iterNum, alpha)
end_time <- Sys.time()
end_time - start_time

# Test result
res.ANCOM_BC=data.frame(genus=rownames(out$feature.table), out$res, 
                        struc.zero[rownames(out$feature.table), ], 
                        row.names = NULL, stringsAsFactors=FALSE)
res.ANCOM_BC$phylum=sapply(res.ANCOM_BC$genus, function(x) strsplit(x, ".g")[[1]][1])
res.ANCOM_BC=res.ANCOM_BC%>%transmute(phylum, genus,
                                      `mean.difference (MA - VEN)` = mean.difference..MA...VEN.,
                                      se, `effect.size (MA - VEN)`=effect.size..MA...VEN., 
                                      p.val, q.val, diff.abn, 
                                      struc.zero.MA=Structural.Zero.in.MA, 
                                      struc.zero.VEN=Structural.Zero.in.VEN)
res.ANCOM_BC=res.ANCOM_BC%>%arrange(q.val)
res.ANCOM_BC.s0=filter(res.ANCOM_BC, genus%in%s0)

write.csv(res.ANCOM_BC, "Output/Real ../data/Genus Level/Taxa_MA vs VEN/MA vs VEN_ANCOM-BC_genus_level.csv", row.names = F)
```

## 6.4 ANCOM

```{r, message=FALSE, warning=FALSE, comment=NA, eval=FALSE}
# Run ANCOM
otu.data=t(obs.abn[s0, ])
otu.data=data.frame(Sample.ID=colnames(obs.abn), otu.data, stringsAsFactors=F)
colnames(otu.data)=c("Sample.ID", s0)

start_time <- Sys.time()
res.W=ANCOM.main(OTUdat=otu.data, Vardat=meta.data, adjusted=FALSE, repeated=F,
                 main.var="country", adj.formula=NULL, repeat.var=NULL, longitudinal=FALSE,
                 random.formula=NULL, multcorr=2, sig=0.05, prev.cut=1.01)

end_time <- Sys.time()
end_time - start_time

res.ANCOM=data.frame(otu.names=c(s0, s1), W_stat=Inf, detected_0.9=TRUE,
                     detected_0.8=TRUE, detected_0.7=TRUE, detected_0.6=TRUE)
res.ANCOM[match(res.W$W.taxa$otu.names, res.ANCOM$otu.names), ]=res.W$W.taxa
res.ANCOM=res.ANCOM%>%arrange(desc(detected_0.7))
res.ANCOM.s0=res.ANCOM[match(s0, res.ANCOM$otu.names), ]

# Tabulation
# res.compare=cbind(select(res.ANCOM_BC.s0, genus, diff.abn), W=res.ANCOM.s0$detected_0.7)
res.compare=left_join(res.ANCOM_BC, res.ANCOM,  by = c("genus" = "otu.names"))
tab.ma.ven=xtabs(~ diff.abn + detected_0.7, data=res.compare)
tab.ma.ven=addmargins(tab.ma.ven)
pander(tab.ma.ven)

write.csv(res.ANCOM, "Output/Real ../data/Genus Level/Taxa_MA vs VEN/MA vs VEN_ANCOM_genus_level.csv", row.names = F)
```

## 6.5 Venn diagram

```{r, message=FALSE, warning=FALSE, comment=NA}
res.ANCOM_BC = read_csv("../data/global_gut/ma_ven_ancom_bc_genus.csv")
res.ANCOM = read_csv("../data/global_gut/ma_ven_ancom_genus.csv.csv")
res.DR = read_tsv("../data/global_gut/ma_ven_dr_genus.tsv")
res.DR = res.DR%>%arrange(desc(`country[T.VEN]`))
# Tabulation
sig.taxa = data.frame(ANCOM_BC = rownames(obs.abn)%in%res.ANCOM_BC[res.ANCOM_BC$diff.abn, ]$genus*1,
                      ANCOM = rownames(obs.abn)%in%res.ANCOM[res.ANCOM$detected_0.7, ]$otu.names*1,
                      DR = rownames(obs.abn)%in%res.DR[c(1:25, (nrow(res.DR)-24):nrow(res.DR)), ]$X1*1)
rownames(sig.taxa) = rownames(obs.abn)
venn.ma_ven = vennCounts(sig.taxa)

vennDiagram(venn.ma_ven, circle.col = c("red", "blue", "green3"), 
            names = c("ANCOM-BC", "ANCOM", "DR"))
```

## 6.6 Normalization efficacy

### 6.61 ANCOM-BC

```{r, message=FALSE, warning=FALSE, comment=NA}
# Paras for ANCOM-BC
grp.name=group.name; grp.ind=group.ind; adj.method="bonferroni"
tol.EM=1e-5; max.iterNum=100; alpha=0.05

# Run ANCOM-BC
out=ANCOM_BC(feature.table, grp.name, grp.ind, struc.zero, adj.method, 
             tol.EM, max.iterNum, alpha)
ancom.bc=out$d+rep(c(out$bias.est, 0), sapply(group.ind, length))

y.ancom.bc=t(t(log(out$feature.table+1))-ancom.bc)
y.ancom.bc=y.ancom.bc+abs(min(y.ancom.bc, na.rm = T))
dist.ancom.bc=vegdist(t(y.ancom.bc), method="bray", na.rm = T) 
fit.ancom.bc=isoMDS(dist.ancom.bc, k=2)
mds.xy.ancom.bc=data.frame(method = "ANCOM-BC", fit.ancom.bc$points)
mds.xy.ancom.bc$country=meta.data$country[match(colnames(out$feature.table), meta.data$Sample.ID)]
```

### 6.62 CSS

```{r, message=FALSE, warning=FALSE, comment=NA}
countdata=obs.abn
zero.threshold=0.90
taxa.info.ind=apply(countdata, 1, function(x) sum(x==0)/ncol(countdata))
countdata=countdata[which(taxa.info.ind<zero.threshold), ]+1L

# CSS: Cumulative-sum scaling
rownames(meta.data)=meta.data$Sample.ID
phenotypeData = Biobase::AnnotatedDataFrame(meta.data)
obj = metagenomeSeq::newMRexperiment(countdata, phenoData=phenotypeData, featureData=NULL)
# Calculating normalization factors
obj = metagenomeSeq::cumNorm(obj)
CSS = metagenomeSeq::normFactors(obj)

y.css=t(t(log(countdata))-log(CSS))
y.css=y.css+abs(min(y.css))
dist.css=vegdist(t(y.css), method="bray", na.rm = T) 
fit.css=isoMDS(dist.css, k=2)
mds.xy.css=data.frame(method = "CSS", fit.css$points, country = meta.data$country)
```

### 6.63 MED

```{r, message=FALSE, warning=FALSE, comment=NA}
coldata=meta.data
count.table=DESeq2::DESeqDataSetFromMatrix(countData = countdata, 
                                           colData = coldata, design = ~ country)
dds=DESeq2::DESeq(count.table, quiet = TRUE)
MED=DESeq2::sizeFactors(dds)

y.med=t(t(log(countdata))-log(MED))
y.med=y.med+abs(min(y.med))
dist.med=vegdist(t(y.med), method="bray", na.rm = T) 
fit.med=isoMDS(dist.med, k=2)
mds.xy.med=data.frame(method = "MED", fit.med$points, country = meta.data$country)
```

### 6.64 UQ

```{r, message=FALSE, warning=FALSE, comment=NA}
groupdata=meta.data$country
dds=edgeR::DGEList(counts = countdata, group = groupdata)
dds=edgeR::calcNormFactors(dds, method="upperquartile")
UQ=dds$samples$norm.factors

y.uq=t(t(log(countdata))-log(UQ))
y.uq=y.uq+abs(min(y.uq))
dist.uq=vegdist(t(y.uq), method="bray", na.rm = T) 
fit.uq=isoMDS(dist.uq, k=2)
mds.xy.uq=data.frame(method = "UQ", fit.uq$points, country = meta.data$country)
```

### 6.65 TMM

```{r, message=FALSE, warning=FALSE, comment=NA}
dds=edgeR::DGEList(counts = countdata, group = groupdata)
dds=edgeR::calcNormFactors(dds, method="TMM")
TMM=dds$samples$norm.factors

y.tmm=t(t(log(countdata))-log(TMM))
y.tmm=y.tmm+abs(min(y.tmm))
dist.tmm=vegdist(t(y.tmm), method="bray", na.rm = T) 
fit.tmm=isoMDS(dist.tmm, k=2)
mds.xy.tmm=data.frame(method = "TMM", fit.tmm$points, country = meta.data$country)
```

### 6.66 TSS

```{r, message=FALSE, warning=FALSE, comment=NA}
TSS=colSums(countdata, na.rm = T)

y.tss=t(t(log(countdata))-log(TSS))
y.tss=y.tss+abs(min(y.tss))
dist.tss=vegdist(t(y.tss), method="bray", na.rm = T) 
fit.tss=isoMDS(dist.tss, k=2)
mds.xy.tss=data.frame(method = "TSS", fit.tss$points, country = meta.data$country)
```

# 7. Fig. 4a

```{r, message=FALSE, warning=FALSE, comment=NA}
p.df = rbind(mds.xy.ancom.bc, mds.xy.css, mds.xy.med, mds.xy.uq, mds.xy.tmm, mds.xy.tss)
p=ggplot(p.df, aes(X1, X2, color = country)) + geom_point(size = 0.2) + 
  facet_wrap(.~method, nrow = 2)+
  labs(x="First NMDS coordinate", y="Second NMDS coordinate")+
  scale_color_discrete(name="", breaks=c("MA", "VEN"), labels = c("Malawi", "Venezuela"))+
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.position = "bottom",
        strip.background = element_rect(fill="white"))+
  guides(color = guide_legend(override.aes = list(size=3)))
ggarrange(p, labels = "a")
ggsave("../figures/Figure 4a.pdf", width=6.25, height=5, units='in')
ggsave("../figures/Figure 4a.jpeg", width=6.25, height=5, units='in', dpi = 300)
```

# 8. Fig. 4b

```{r, message=FALSE, warning=FALSE, comment=NA}
class(venn.ma_us) = "matrix"; class(venn.ven_us) = "matrix"; class(venn.ma_ven) = "matrix"
df.venn = data.frame(x = c(0, 0.866, -0.866),
                     y = c(1, -0.5, -0.5),
                     labels = c("ANCOM-BC", "DR", "ANCOM"))

df.ma_us = as.data.frame(venn.ma_us)%>%
  mutate(x = c(-2, 1.2, -1.2, 0, 0, 0.8, -0.8, 0),
         y = c(2, -0.6, -0.6, -1, 1.2, 0.5, 0.5, 0))
df.ven_us = as.data.frame(venn.ven_us)%>%
  mutate(x = c(-2, 1.2, -1.2, 0, 0, 0.8, -0.8, 0),
         y = c(2, -0.6, -0.6, -1, 1.2, 0.5, 0.5, 0))
df.ma_ven = as.data.frame(venn.ma_ven)%>%
  mutate(x = c(-2, 1.2, -1.2, 0, 0, 0.8, -0.8, 0),
         y = c(2, -0.6, -0.6, -1, 1.2, 0.5, 0.5, 0))

df.venn = rbind(data.frame(df.venn, compare = "MA vs US"), 
                data.frame(df.venn, compare = "VEN vs US"), 
                data.frame(df.venn, compare = "MA vs VEN"))
df.venn$labels = factor(df.venn$labels, levels = c("ANCOM-BC", "ANCOM", "DR"))
df.vdc = rbind(data.frame(df.ma_us, compare = "MA vs US"),
               data.frame(df.ven_us, compare = "VEN vs US"),
               data.frame(df.ma_ven, compare = "MA vs VEN"))%>%
  transmute(x, y, label = Counts, compare)

p = ggplot(df.venn) +
  geom_circle(aes(x0 = x, y0 = y, r = 1.5, fill = labels), alpha = 0.5, size = 1, colour = 'grey') +
  coord_fixed() + facet_wrap(.~compare, nrow = 1)+
  labs(x = NULL, y = NULL, fill = NULL) +
  geom_text(data = df.vdc, aes(x, y, label=label), inherit.aes=FALSE, size = 5)+
  theme_bw()+
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.position = "bottom",
        strip.background = element_rect(fill="white"),
        axis.text = element_blank(),
        axis.ticks = element_blank())
ggarrange(p, labels = "b")
ggsave("../figures/Figure 4b.pdf", width=6.25, height=5, units='in')
ggsave("../figures/Figure 4b.jpeg", width=6.25, height=5, units='in', dpi = 300)
```

# Session information

```{r, message=FALSE, warning=FALSE, comment=NA}
sessionInfo()
devtools::session_info()
```


