---
title: "Figure 4"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
rm(list = ls())

library(tidyverse)
library(nloptr)
library(phyloseq)
library(stringr)
library(RColorBrewer)
library(ggpubr)
library(ggforce)
library(limma)
library(vegan)
library(MASS)
library(pander)
panderOptions('table.caption.prefix', NULL)
panderOptions('table.continues', NULL)
panderOptions('table.emphasize.rownames', FALSE)

source("ancom_bc_v1.0.R")
source("ancom_v1.0.R")
```

# 1. Read in OTU table and meta data

```{r, message=FALSE, warning=FALSE, comment=NA}
# Meta data
meta_data=read_tsv("../data/global_gut/global_gut_metadata.txt")
meta_data=meta_data%>%transmute(Sample.ID=`#SampleID`, age=AGE, sex=SEX, country=COUNTRY)%>%
  arrange(Sample.ID)
meta_data=meta_data[complete.cases(meta_data), ]
meta_data$age=as.numeric(meta_data$age)
meta_data$country=recode(meta_data$country, `GAZ:Malawi` = "MA", 
                         `GAZ:United States of America` = "US", `GAZ:Venezuela` = "VEN")

# Taxonomy
taxonomy=read_tsv("../data/global_gut/global_gut_taxonomy.txt")
taxonomy=taxonomy%>%rowwise()%>%
  mutate(genus_name=paste(Phylum, Genus, sep = ";"))

# OTU table
otu_table=read_tsv("../data/global_gut/global_gut_otu.txt")
otu_table=otu_table[, -532]
otu_table$OTU_ID=taxonomy$genus_name[match(otu_table$OTU_ID, taxonomy$OTU_ID)]
otu_table=as.data.frame(otu_table)
otu_table[, -1]=apply(otu_table[, -1], 2, as.numeric)

# Aggregate into genus level
genus_table=otu_table%>%group_by(OTU_ID)%>%
  summarise_all(sum)
non_info_pos=grep("\\g__\\b", genus_table$OTU_ID) # Exact match
genus_table=genus_table[-non_info_pos, ]
genus_table=as.data.frame(genus_table)
```

# 2. MA vs US

## 2.1 Subset OTU table and meta data

```{r, message=FALSE, warning=FALSE, comment=NA}
# Subset meta data
meta.data=meta_data%>%filter(country%in%c("MA", "US"))
meta.data$country=as.character(meta.data$country)
# Subset OTU table
obs.abn=genus_table
rownames(obs.abn)=obs.abn$OTU_ID
obs.abn=obs.abn[, -1]
obs.abn=obs.abn[, meta.data$Sample.ID]

# Recode sample ID
meta.data$Sample.ID=seq(nrow(meta.data))
colnames(obs.abn)=seq(nrow(meta.data))

write_tsv(meta.data, "../data/songbird/meta_ma_us.txt")
write_tsv(obs.abn%>%rownames_to_column("genus"), "../data/songbird/otu_ma_us.txt")
```

## 2.2 Pre-Processing

```{r, message=FALSE, warning=FALSE, comment=NA}
feature.table=obs.abn; sample.var="Sample.ID"; group.var="country"; 
zero.cut=0.90; lib.cut=1000; neg.lb=TRUE
pre.process=feature_table_pre_process(feature.table, meta.data, sample.var, 
                                      group.var, zero.cut, lib.cut, neg.lb)
feature.table=pre.process$feature.table
library.size=pre.process$library.size
group.name=pre.process$group.name
group.ind=pre.process$group.ind
struc.zero=pre.process$structure.zeros
num.struc.zero=apply(struc.zero, 1, sum)
s0=rownames(feature.table)[which(num.struc.zero==0)]
s1=rownames(feature.table)[which(num.struc.zero==1)]
```

## 2.3 ANCOM-BC

```{r, message=FALSE, warning=FALSE, comment=NA, eval=FALSE}
# Paras for ANCOM-BC
grp.name=group.name; grp.ind=group.ind; adj.method="bonferroni"
tol.EM=1e-5; max.iterNum=100; perNum=1000; alpha=0.05

# Run ANCOM-BC
start_time <- Sys.time()
out=ANCOM_BC(feature.table, grp.name, grp.ind, struc.zero, adj.method, 
              tol.EM, max.iterNum, perNum, alpha)
end_time <- Sys.time()
end_time - start_time

# Test result
res.ANCOM_BC=data.frame(genus=rownames(out$feature.table), out$res, 
                        struc.zero[rownames(out$feature.table), ], 
                        row.names = NULL, stringsAsFactors=FALSE, check.names = FALSE)
res.ANCOM_BC$phylum=sapply(res.ANCOM_BC$genus, function(x) strsplit(x, ".g")[[1]][1])
res.ANCOM_BC=res.ANCOM_BC%>%transmute(phylum, genus,
                                      `mean.difference (MA - US)` = -`mean.difference (US - MA)`,
                                      `se (MA - US)` = `se (US - MA)`, 
                                      `effect.size (MA - US)` = ifelse(is.infinite(W),  Inf, -W), 
                                      p.val, q.val, diff.abn, 
                                      `structural.zero (MA)`, `structural.zero (US)`)
res.ANCOM_BC=res.ANCOM_BC%>%arrange(q.val)
res.ANCOM_BC.s0=filter(res.ANCOM_BC, genus%in%s0)

write.csv(res.ANCOM_BC, "../data/global_gut/ma_us_ancom_bc_genus.csv", row.names = F)
```

## 2.4 ANCOM

```{r, message=FALSE, warning=FALSE, comment=NA, eval=FALSE}
# Run ANCOM
otu.data=t(obs.abn[s0, ])
otu.data=data.frame(Sample.ID=colnames(obs.abn), otu.data, stringsAsFactors=F)
colnames(otu.data)=c("Sample.ID", s0)

start_time <- Sys.time()
res.W=ANCOM.main(OTUdat=otu.data, Vardat=meta.data, adjusted=FALSE, repeated=FALSE,
                 main.var="country", adj.formula=NULL, repeat.var=NULL, longitudinal=FALSE,
                 random.formula=NULL, multcorr=2, sig=0.05, prev.cut=1.01)

end_time <- Sys.time()
end_time - start_time

res.ANCOM=data.frame(otu.names=c(s0, s1), W_stat=Inf, detected_0.9=TRUE,
                     detected_0.8=TRUE, detected_0.7=TRUE, detected_0.6=TRUE)
res.ANCOM[match(res.W$W.taxa$otu.names, res.ANCOM$otu.names), ]=res.W$W.taxa
res.ANCOM=res.ANCOM%>%arrange(desc(detected_0.7))
res.ANCOM.s0=res.ANCOM[match(s0, res.ANCOM$otu.names), ]

write.csv(res.ANCOM, "../data/global_gut/ma_us_ancom_genus.csv", row.names = F)
```

## 2.5 Venn diagram

```{r, message=FALSE, warning=FALSE, comment=NA}
res.ANCOM_BC = read_csv("../data/global_gut/ma_us_ancom_bc_genus.csv")
res.ANCOM = read_csv("../data/global_gut/ma_us_ancom_genus.csv")
res.DR = read_tsv("../data/global_gut/ma_us_dr_genus.tsv")
res.DR = res.DR%>%arrange(desc(`country[T.US]`))
# Tabulation
sig.taxa = data.frame(ANCOM_BC = rownames(obs.abn)%in%res.ANCOM_BC[res.ANCOM_BC$diff.abn, ]$genus*1,
                      ANCOM = rownames(obs.abn)%in%res.ANCOM[res.ANCOM$detected_0.7, ]$otu.names*1,
                      DR = rownames(obs.abn)%in%res.DR[c(1:25, (nrow(res.DR)-24):nrow(res.DR)), ]$X1*1)
rownames(sig.taxa) = rownames(obs.abn)
venn.ma_us = vennCounts(sig.taxa)

vennDiagram(venn.ma_us, circle.col = c("red", "blue", "green3"), 
            names = c("ANCOM-BC", "ANCOM", "DR"))
```

# 3. VEN vs US

## 3.1 Subset OTU table and meta data

```{r, message=FALSE, warning=FALSE, comment=NA}
# Subset meta data
meta.data=meta_data%>%filter(country%in%c("VEN", "US"))
meta.data$country=as.character(meta.data$country)
# Subset OTU table
obs.abn=genus_table
rownames(obs.abn)=obs.abn$OTU_ID
obs.abn=obs.abn[, -1]
obs.abn=obs.abn[, meta.data$Sample.ID]

# Recode sample ID
meta.data$Sample.ID=seq(nrow(meta.data))
colnames(obs.abn)=seq(nrow(meta.data))

write_tsv(meta.data, "../data/songbird/meta_ven_us.txt")
write_tsv(obs.abn%>%rownames_to_column("genus"), "../data/songbird/otu_ven_us.txt")
```

## 3.2 Pre-Processing

```{r, message=FALSE, warning=FALSE, comment=NA}
feature.table=obs.abn; sample.var="Sample.ID"; group.var="country"; 
zero.cut=0.90; lib.cut=1000; neg.lb=TRUE
pre.process=feature_table_pre_process(feature.table, meta.data, sample.var, 
                                      group.var, zero.cut, lib.cut, neg.lb)
feature.table=pre.process$feature.table
library.size=pre.process$library.size
group.name=pre.process$group.name
group.ind=pre.process$group.ind
struc.zero=pre.process$structure.zeros
num.struc.zero=apply(struc.zero, 1, sum)
s0=rownames(feature.table)[which(num.struc.zero==0)]
s1=rownames(feature.table)[which(num.struc.zero==1)]
```

## 3.3 ANCOM-BC

```{r, message=FALSE, warning=FALSE, comment=NA, eval=FALSE}
# Paras for ANCOM-BC
grp.name=group.name; grp.ind=group.ind; adj.method="bonferroni"
tol.EM=1e-5; max.iterNum=100; perNum=1000; alpha=0.05

# Run ANCOM-BC
start_time <- Sys.time()
out=ANCOM_BC(feature.table, grp.name, grp.ind, struc.zero, adj.method, 
              tol.EM, max.iterNum, perNum, alpha)
end_time <- Sys.time()
end_time - start_time

# Test result
res.ANCOM_BC=data.frame(genus=rownames(out$feature.table), out$res, 
                        struc.zero[rownames(out$feature.table), ], 
                        row.names = NULL, stringsAsFactors=FALSE, check.names = FALSE)
res.ANCOM_BC$phylum=sapply(res.ANCOM_BC$genus, function(x) strsplit(x, ".g")[[1]][1])
res.ANCOM_BC=res.ANCOM_BC%>%dplyr::select(phylum, everything())
res.ANCOM_BC=res.ANCOM_BC%>%arrange(q.val)
res.ANCOM_BC.s0=filter(res.ANCOM_BC, genus%in%s0)

write.csv(res.ANCOM_BC, "../data/global_gut/ven_us_ancom_bc_genus.csv", row.names = F)
```

## 3.4 ANCOM

```{r, message=FALSE, warning=FALSE, comment=NA, eval=FALSE}
# Run ANCOM
otu.data=t(obs.abn[s0, ])
otu.data=data.frame(Sample.ID=colnames(obs.abn), otu.data, stringsAsFactors=F)
colnames(otu.data)=c("Sample.ID", s0)

start_time <- Sys.time()
res.W=ANCOM.main(OTUdat=otu.data, Vardat=meta.data, adjusted=FALSE, repeated=F,
                 main.var="country", adj.formula=NULL, repeat.var=NULL, longitudinal=FALSE,
                 random.formula=NULL, multcorr=2, sig=0.05, prev.cut=1.01)

end_time <- Sys.time()
end_time - start_time

res.ANCOM=data.frame(otu.names=c(s0, s1), W_stat=Inf, detected_0.9=TRUE,
                     detected_0.8=TRUE, detected_0.7=TRUE, detected_0.6=TRUE)
res.ANCOM[match(res.W$W.taxa$otu.names, res.ANCOM$otu.names), ]=res.W$W.taxa
res.ANCOM=res.ANCOM%>%arrange(desc(detected_0.7))
res.ANCOM.s0=res.ANCOM[match(s0, res.ANCOM$otu.names), ]

# Tabulation
res.compare=left_join(res.ANCOM_BC, res.ANCOM,  by = c("genus" = "otu.names"))
tab.ven.us=xtabs(~ diff.abn + detected_0.7, data=res.compare)
tab.ven.us=addmargins(tab.ven.us)
pander(tab.ven.us)

write.csv(res.ANCOM, "../data/global_gut/ven_us_ancom_genus.cs", row.names = F)
```

## 3.5 Venn diagram

```{r, message=FALSE, warning=FALSE, comment=NA}
res.ANCOM_BC = read_csv("../data/global_gut/ven_us_ancom_bc_genus.csv")
res.ANCOM = read_csv("../data/global_gut/ven_us_ancom_genus.csv")
res.DR = read_tsv("../data/global_gut/ven_us_dr_genus.tsv")
res.DR = res.DR%>%arrange(desc(`country[T.VEN]`))
# Tabulation
sig.taxa = data.frame(ANCOM_BC = rownames(obs.abn)%in%res.ANCOM_BC[res.ANCOM_BC$diff.abn, ]$genus*1,
                      ANCOM = rownames(obs.abn)%in%res.ANCOM[res.ANCOM$detected_0.7, ]$otu.names*1,
                      DR = rownames(obs.abn)%in%res.DR[c(1:25, (nrow(res.DR)-24):nrow(res.DR)), ]$X1*1)
rownames(sig.taxa) = rownames(obs.abn)
venn.ven_us = vennCounts(sig.taxa)

vennDiagram(venn.ven_us, circle.col = c("red", "blue", "green3"), 
            names = c("ANCOM-BC", "ANCOM", "DR"))
```

# 4. MA vs VEN

## 4.1 Subset OTU table and meta data

```{r, message=FALSE, warning=FALSE, comment=NA}
# Subset meta data
meta.data=meta_data%>%filter(country%in%c("MA", "VEN"))
meta.data$country=as.character(meta.data$country)
# Subset OTU table
obs.abn=genus_table
rownames(obs.abn)=obs.abn$OTU_ID
obs.abn=obs.abn[, -1]
obs.abn=obs.abn[, meta.data$Sample.ID]

# Recode sample ID
meta.data$Sample.ID=seq(nrow(meta.data))
colnames(obs.abn)=seq(nrow(meta.data))

write_tsv(meta.data, "../data/songbird/meta_ma_ven.txt")
write_tsv(obs.abn%>%rownames_to_column("genus"), "../data/songbird/otu_ma_ven.txt")
```

## 4.2 Pre-Processing

```{r, message=FALSE, warning=FALSE, comment=NA}
feature.table=obs.abn; sample.var="Sample.ID"; group.var="country"; 
zero.cut=0.90; lib.cut=1000; neg.lb=TRUE
pre.process=feature_table_pre_process(feature.table, meta.data, sample.var, 
                                      group.var, zero.cut, lib.cut, neg.lb)
feature.table=pre.process$feature.table
library.size=pre.process$library.size
group.name=pre.process$group.name
group.ind=pre.process$group.ind
struc.zero=pre.process$structure.zeros
num.struc.zero=apply(struc.zero, 1, sum)
s0=rownames(feature.table)[which(num.struc.zero==0)]
s1=rownames(feature.table)[which(num.struc.zero==1)]
```

## 4.3 ANCOM-BC

```{r, message=FALSE, warning=FALSE, comment=NA, eval=FALSE}
# Paras for ANCOM-BC
grp.name=group.name; grp.ind=group.ind; adj.method="bonferroni"
tol.EM=1e-5; max.iterNum=100; perNum=1000; alpha=0.05

# Run ANCOM-BC
start_time <- Sys.time()
out=ANCOM_BC(feature.table, grp.name, grp.ind, struc.zero, adj.method, 
              tol.EM, max.iterNum, perNum, alpha)
end_time <- Sys.time()
end_time - start_time

# Test result
res.ANCOM_BC=data.frame(genus=rownames(out$feature.table), out$res, 
                        struc.zero[rownames(out$feature.table), ], 
                        row.names = NULL, stringsAsFactors=FALSE, check.names = FALSE)
res.ANCOM_BC$phylum=sapply(res.ANCOM_BC$genus, function(x) strsplit(x, ".g")[[1]][1])
res.ANCOM_BC=res.ANCOM_BC%>%transmute(phylum, genus,
                                      `mean.difference (MA - VEN)` = -`mean.difference (VEN - MA)`,
                                      `se (MA - VEN)` = `se (VEN - MA)`, 
                                      `effect.size (MA - VEN)` = ifelse(is.infinite(W),  Inf, -W), 
                                      p.val, q.val, diff.abn, 
                                      `structural.zero (MA)`, `structural.zero (VEN)`)
res.ANCOM_BC=res.ANCOM_BC%>%arrange(q.val)
res.ANCOM_BC.s0=filter(res.ANCOM_BC, genus%in%s0)

write.csv(res.ANCOM_BC, "../data/global_gut/ma_ven_ancom_bc_genus.csv", row.names = F)
```

## 4.4 ANCOM

```{r, message=FALSE, warning=FALSE, comment=NA, eval=FALSE}
# Run ANCOM
otu.data=t(obs.abn[s0, ])
otu.data=data.frame(Sample.ID=colnames(obs.abn), otu.data, stringsAsFactors=F)
colnames(otu.data)=c("Sample.ID", s0)

start_time <- Sys.time()
res.W=ANCOM.main(OTUdat=otu.data, Vardat=meta.data, adjusted=FALSE, repeated=F,
                 main.var="country", adj.formula=NULL, repeat.var=NULL, longitudinal=FALSE,
                 random.formula=NULL, multcorr=2, sig=0.05, prev.cut=1.01)

end_time <- Sys.time()
end_time - start_time

res.ANCOM=data.frame(otu.names=c(s0, s1), W_stat=Inf, detected_0.9=TRUE,
                     detected_0.8=TRUE, detected_0.7=TRUE, detected_0.6=TRUE)
res.ANCOM[match(res.W$W.taxa$otu.names, res.ANCOM$otu.names), ]=res.W$W.taxa
res.ANCOM=res.ANCOM%>%arrange(desc(detected_0.7))
res.ANCOM.s0=res.ANCOM[match(s0, res.ANCOM$otu.names), ]

# Tabulation
res.compare=left_join(res.ANCOM_BC, res.ANCOM,  by = c("genus" = "otu.names"))
tab.ma.ven=xtabs(~ diff.abn + detected_0.7, data=res.compare)
tab.ma.ven=addmargins(tab.ma.ven)
pander(tab.ma.ven)

write.csv(res.ANCOM, "../data/global_gut/ma_ven_ancom_genus.csv", row.names = F)
```

## 4.5 Venn diagram

```{r, message=FALSE, warning=FALSE, comment=NA}
res.ANCOM_BC = read_csv("../data/global_gut/ma_ven_ancom_bc_genus.csv")
res.ANCOM = read_csv("../data/global_gut/ma_ven_ancom_genus.csv")
res.DR = read_tsv("../data/global_gut/ma_ven_dr_genus.tsv")
res.DR = res.DR%>%arrange(desc(`country[T.VEN]`))
# Tabulation
sig.taxa = data.frame(ANCOM_BC = rownames(obs.abn)%in%res.ANCOM_BC[res.ANCOM_BC$diff.abn, ]$genus*1,
                      ANCOM = rownames(obs.abn)%in%res.ANCOM[res.ANCOM$detected_0.7, ]$otu.names*1,
                      DR = rownames(obs.abn)%in%res.DR[c(1:25, (nrow(res.DR)-24):nrow(res.DR)), ]$X1*1)
rownames(sig.taxa) = rownames(obs.abn)
venn.ma_ven = vennCounts(sig.taxa)

vennDiagram(venn.ma_ven, circle.col = c("red", "blue", "green3"), 
            names = c("ANCOM-BC", "ANCOM", "DR"))
```

## 4.6 Normalization efficacy

### 4.61 ANCOM-BC

```{r, message=FALSE, warning=FALSE, comment=NA}
# Paras for ANCOM-BC
grp.name=group.name; grp.ind=group.ind; adj.method="bonferroni"
tol.EM=1e-5; max.iterNum=100; perNum=1000; alpha=0.05

# Run ANCOM-BC
out=ANCOM_BC(feature.table, grp.name, grp.ind, struc.zero, adj.method, 
             tol.EM, max.iterNum, perNum, alpha)
ancom.bc=out$d

y.ancom.bc=t(t(log(out$feature.table+1))-ancom.bc)
y.ancom.bc=y.ancom.bc+abs(min(y.ancom.bc, na.rm = T))
dist.ancom.bc=vegdist(t(y.ancom.bc), method="bray", na.rm = T) 
fit.ancom.bc=isoMDS(dist.ancom.bc, k=2)
mds.xy.ancom.bc=data.frame(method = "ANCOM-BC", fit.ancom.bc$points)
mds.xy.ancom.bc$country=meta.data$country[match(colnames(out$feature.table), meta.data$Sample.ID)]
```

### 4.62 CSS

```{r, message=FALSE, warning=FALSE, comment=NA}
countdata=obs.abn
zero.threshold=0.90
taxa.info.ind=apply(countdata, 1, function(x) sum(x==0)/ncol(countdata))
countdata=countdata[which(taxa.info.ind<zero.threshold), ]+1L

# CSS: Cumulative-sum scaling
rownames(meta.data)=meta.data$Sample.ID
phenotypeData = Biobase::AnnotatedDataFrame(meta.data)
obj = metagenomeSeq::newMRexperiment(countdata, phenoData=phenotypeData, featureData=NULL)
# Calculating normalization factors
obj = metagenomeSeq::cumNorm(obj)
CSS = metagenomeSeq::normFactors(obj)

y.css=t(t(log(countdata))-log(CSS))
y.css=y.css+abs(min(y.css))
dist.css=vegdist(t(y.css), method="bray", na.rm = T) 
fit.css=isoMDS(dist.css, k=2)
mds.xy.css=data.frame(method = "CSS", fit.css$points, country = meta.data$country)
```

### 4.63 MED

```{r, message=FALSE, warning=FALSE, comment=NA}
coldata=meta.data
count.table=DESeq2::DESeqDataSetFromMatrix(countData = countdata, 
                                           colData = coldata, design = ~ country)
dds=DESeq2::DESeq(count.table, quiet = TRUE)
MED=DESeq2::sizeFactors(dds)

y.med=t(t(log(countdata))-log(MED))
y.med=y.med+abs(min(y.med))
dist.med=vegdist(t(y.med), method="bray", na.rm = T) 
fit.med=isoMDS(dist.med, k=2)
mds.xy.med=data.frame(method = "MED", fit.med$points, country = meta.data$country)
```

### 4.64 UQ

```{r, message=FALSE, warning=FALSE, comment=NA}
groupdata=meta.data$country
dds=edgeR::DGEList(counts = countdata, group = groupdata)
dds=edgeR::calcNormFactors(dds, method="upperquartile")
UQ=dds$samples$norm.factors

y.uq=t(t(log(countdata))-log(UQ))
y.uq=y.uq+abs(min(y.uq))
dist.uq=vegdist(t(y.uq), method="bray", na.rm = T) 
fit.uq=isoMDS(dist.uq, k=2)
mds.xy.uq=data.frame(method = "UQ", fit.uq$points, country = meta.data$country)
```

### 4.65 TMM

```{r, message=FALSE, warning=FALSE, comment=NA}
dds=edgeR::DGEList(counts = countdata, group = groupdata)
dds=edgeR::calcNormFactors(dds, method="TMM")
TMM=dds$samples$norm.factors

y.tmm=t(t(log(countdata))-log(TMM))
y.tmm=y.tmm+abs(min(y.tmm))
dist.tmm=vegdist(t(y.tmm), method="bray", na.rm = T) 
fit.tmm=isoMDS(dist.tmm, k=2)
mds.xy.tmm=data.frame(method = "TMM", fit.tmm$points, country = meta.data$country)
```

### 4.66 TSS

```{r, message=FALSE, warning=FALSE, comment=NA}
TSS=colSums(countdata, na.rm = T)

y.tss=t(t(log(countdata))-log(TSS))
y.tss=y.tss+abs(min(y.tss))
dist.tss=vegdist(t(y.tss), method="bray", na.rm = T) 
fit.tss=isoMDS(dist.tss, k=2)
mds.xy.tss=data.frame(method = "TSS", fit.tss$points, country = meta.data$country)
```

## 4.7 Between Sum of Squares

```{r, message=FALSE, warning=FALSE, comment=NA}
ss_compute=function(x){
  x=as.matrix(x)
  ss=sum(t(t(x)-colMeans(x, na.rm = T))^2, na.rm = T)
  return(ss)
}

mds.xy.ancom.bc=mds.xy.ancom.bc%>%mutate(scale_x1=scale(X1), scale_x2=scale(X2))
mds.xy.css=mds.xy.css%>%mutate(scale_x1=scale(X1), scale_x2=scale(X2))
mds.xy.med=mds.xy.med%>%mutate(scale_x1=scale(X1), scale_x2=scale(X2))
mds.xy.uq=mds.xy.uq%>%mutate(scale_x1=scale(X1), scale_x2=scale(X2))
mds.xy.tmm=mds.xy.tmm%>%mutate(scale_x1=scale(X1), scale_x2=scale(X2))
mds.xy.tss=mds.xy.tss%>%mutate(scale_x1=scale(X1), scale_x2=scale(X2))

mds.list=list(mds.xy.ancom.bc, mds.xy.css, mds.xy.med, mds.xy.uq, mds.xy.tmm, mds.xy.tss)
ss.ratio.df=data.frame(method=c("ANCOM-BC", "CSS", "MED", "UQ", "TMM", "TSS"), ss.ratio = NA)
for (i in 1:length(mds.list)) {
  method=mds.list[[i]]
  sst=ss_compute(method%>%dplyr::select(scale_x1, scale_x2))
  ssw=ss_compute(method%>%filter(country=="MA")%>%dplyr::select(scale_x1, scale_x2))+
    ss_compute(method%>%filter(country=="VEN")%>%dplyr::select(scale_x1, scale_x2))
  ssb=sst-ssw
  ss.ratio=ssb/ssw
  ss.ratio.df[i, "ss.ratio"]=ss.ratio
}

mds.list=list(mds.xy.ancom.bc, mds.xy.css, mds.xy.med, mds.xy.uq, mds.xy.tmm, mds.xy.tss)
ss.ratio.df=data.frame(method=c("ANCOM-BC", "CSS", "MED", "UQ", "TMM", "TSS"), ss.ratio = NA)
for (i in 1:length(mds.list)) {
  method=mds.list[[i]]
  sst=ss_compute(method%>%dplyr::select(X1, X2))
  ssw=ss_compute(method%>%filter(country=="MA")%>%dplyr::select(X1, X2))+
    ss_compute(method%>%filter(country=="VEN")%>%dplyr::select(X1, X2))
  ssb=sst-ssw
  ss.ratio=ssb/ssw
  ss.ratio.df[i, "ss.ratio"]=ss.ratio
}
```

# 5. Fig. 4a

```{r, message=FALSE, warning=FALSE, comment=NA}
txt.df = data.frame(X1=0.2, X2=0.25, ssb.df)
txt.df$ssb = paste0("SSB = ", round(txt.df$ssb, 0))
p.df = rbind(mds.xy.ancom.bc, mds.xy.css, mds.xy.med, mds.xy.uq, mds.xy.tmm, mds.xy.tss)
p=ggplot(p.df, aes(X1, X2, color = country)) + geom_point(size = 0.3) + 
  facet_wrap(.~method, nrow = 2)+scale_y_continuous(limits = c(-0.3, 0.3))+
  labs(x="First NMDS coordinate", y="Second NMDS coordinate")+
  scale_color_discrete(name="", breaks=c("MA", "VEN"), labels = c("Malawi", "Venezuela"))+
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.position = "bottom",
        strip.background = element_rect(fill="white"))+
  guides(color = guide_legend(override.aes = list(size=3)))+
  geom_text(data = txt.df, mapping = aes(x = X1, y = X2, label = ssb, color = NULL,group= NULL))
ggarrange(p, labels = "a")
ggsave("../figures/Figure 4a.pdf", width=6.25, height=5, units='in')
ggsave("../figures/Figure 4a.jpeg", width=6.25, height=5, units='in', dpi = 300)
```

# 6. Fig. 4b

```{r, message=FALSE, warning=FALSE, comment=NA}
class(venn.ma_us) = "matrix"; class(venn.ven_us) = "matrix"; class(venn.ma_ven) = "matrix"
df.venn = data.frame(x = c(0, 0.866, -0.866),
                     y = c(1, -0.5, -0.5),
                     labels = c("ANCOM-BC", "DR", "ANCOM"))

df.ma_us = as.data.frame(venn.ma_us)%>%
  mutate(x = c(-2, 1.2, -1.2, 0, 0, 0.8, -0.8, 0),
         y = c(2, -0.6, -0.6, -1, 1.2, 0.5, 0.5, 0))
df.ven_us = as.data.frame(venn.ven_us)%>%
  mutate(x = c(-2, 1.2, -1.2, 0, 0, 0.8, -0.8, 0),
         y = c(2, -0.6, -0.6, -1, 1.2, 0.5, 0.5, 0))
df.ma_ven = as.data.frame(venn.ma_ven)%>%
  mutate(x = c(-2, 1.2, -1.2, 0, 0, 0.8, -0.8, 0),
         y = c(2, -0.6, -0.6, -1, 1.2, 0.5, 0.5, 0))

df.venn = rbind(data.frame(df.venn, compare = "MA vs US"), 
                data.frame(df.venn, compare = "VEN vs US"), 
                data.frame(df.venn, compare = "MA vs VEN"))
df.venn$labels = factor(df.venn$labels, levels = c("ANCOM-BC", "ANCOM", "DR"))
df.vdc = rbind(data.frame(df.ma_us, compare = "MA vs US"),
               data.frame(df.ven_us, compare = "VEN vs US"),
               data.frame(df.ma_ven, compare = "MA vs VEN"))%>%
  transmute(x, y, label = Counts, compare)

p = ggplot(df.venn) +
  geom_circle(aes(x0 = x, y0 = y, r = 1.5, fill = labels), alpha = 0.5, size = 1, colour = 'grey') +
  coord_fixed() + facet_wrap(.~compare, nrow = 1)+
  labs(x = NULL, y = NULL, fill = NULL) +
  geom_text(data = df.vdc, aes(x, y, label=label), inherit.aes=FALSE, size = 5)+
  theme_bw()+
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.position = "bottom",
        strip.background = element_rect(fill="white"),
        axis.text = element_blank(),
        axis.ticks = element_blank())
ggarrange(p, labels = "b")
ggsave("../figures/Figure 4b.pdf", width=6.25, height=5, units='in')
ggsave("../figures/Figure 4b.jpeg", width=6.25, height=5, units='in', dpi = 300)
```

# Session information

```{r, message=FALSE, warning=FALSE, comment=NA}
sessionInfo()
devtools::session_info()
```


